astrosat_users/models/models_customers.py

```diff
@@ -149,10 +153,38 @@ def clean(self):
         if self.pk is None and self.user in self.customer.users.all():
             raise ValidationError("User is already a member of Customer.")
 
-    def invite(self):
-        # TODO: for now I'm just setting the date, need to actually do the invite
-        self.invitation_date = timezone.now()  # (passes a datetime object using the timezone from settings.py)
-        self.save()
+    def invite(self, **kwargs):
+        """
+        Makes the user aware that they are now a member of the customer
+        by sending an invitation email to their registered address.  If
+        they are a newly-created user, they will have to update their
+        password; the email will contain instructions on how to do this.
+        """
+
+        adapter = kwargs.get("adapter", get_adapter())
+        context = kwargs.get("context", {})
+        template_prefix = kwargs.get("template_prefix", "astrosat_users/email/invitation")
 
-# class CustomerInvitation(models.Model):
-#     pass
+        user = self.user
+        customer = self.customer
+        context.update({
+            "user": user,
+            "customer": customer,
+        })
+
+        if user.change_password:
+            token_generator = kwargs.get("token_generator", adapter.default_token_generator)
+            token_key = token_generator.make_token(user)
+            url = adapter.get_password_confirmation_url(adapter.request, user, token_key)
+            context["password_reset_url"] = url
+
+        if (
+            allauth_app_settings.AUTHENTICATION_METHOD
+            != allauth_app_settings.AuthenticationMethod.EMAIL
+        ):
+            context["username"] = user_username(user)
+
+        adapter.send_mail(template_prefix, user.email, context)
```

Comment:
Couple of questions here:

1. Why do we need to use different methods to send the email? the old `adapter.send_password_confirmation_email` and new `adapter.send_mail`
2. Why does more info need to be collected for the `context` in the new code, but not in the old, are you changing the template as well?
3. More a comment on style to see if you would ever do this, but I would collate all the data needed for the context and apply all to the context at once. I'm not looking for any changes, more whether you feel that is a valid alternative?

Reply:
1. `adapter.send_password_confirmation_email` gets run when a user chooses to reset their password.  This is not sending a password reset email, per se.  Rather it is sending an invitation email that includes password reset information.
2.  If you're referring to no longer passing `manager`, it's b/c I realised I wasn't actually using that
3.  I'm not sure what you mean.  The context is built up by a series of `if` blocks.  I only add `password_reset_url` and `username` if they're needed.

Reply:
Okay this all makes sense:

1. I was confused by the name `send_password_confirmation_email` I would expect it to be called `send_password_reset_email` so it is clear, but that's fine, I get it now. Maybe `send_mail` is too generic, but not pushing for it to be changed.
2. That is pretty much what I meant, but I've since notice that `user` is not used in the template, least not that I found.
3. I would always pass in the same set of parameters, but those not to be used would be `null`. I'm not pushing for change, more just explaining what I would do to see if you feel it is a valid approach. I just prefer everything to always be sent to the template (using null) for those with no value, rather than sometimes it gets these params and other times it gets all. As I said before, it is just differing styles.

I'm happy to resolve this conversation, but could you just double-check if `user` is used in the template, as it is passed into the context, but I didn't see it being used.

Reply:
You're right; `user` is not used.

Reply:
re: using `null`

I think this is a javascript vs django-template (not necessarily python) thing.  By default if I request context that is not defined in a template I will get a falsey value.  I'm not sure what happens if it _is_ defined but is `null`.  I suspect it would still work.